pkg/external-provider/series_converter.go

```diff
@@ -0,0 +1,289 @@
+	queryParts := []queryPart{}
+
+	// Build up the namespace part of the query.
+	namespaceQueryPart, err := c.buildNamespaceQueryPartForSeries(namespace)
+	if err != nil {
+		return "", err
+	}
+
+	if namespaceQueryPart.labelName != "" {
+		queryParts = append(queryParts, namespaceQueryPart)
+	}
+
+	// Build up the resource part of the query.
+	resourceQueryPart, err := c.buildResourceQueryPartForSeries(resource, names...)
+	if err != nil {
+		return "", err
+	}
+
+	if resourceQueryPart.labelName != "" {
+		queryParts = append(queryParts, resourceQueryPart)
+	}
+
+	return c.queryBuilder.BuildSelector(series, resourceQueryPart.labelName, []string{resourceQueryPart.labelName}, queryParts)
+}
+
+// ConvertersFromConfig produces a MetricNamer for each rule in the given config.
+func ConvertersFromConfig(cfg *config.MetricsDiscoveryConfig, mapper apimeta.RESTMapper) ([]SeriesConverter, []error) {
+	errs := []error{}
+	converters := []SeriesConverter{}
+
+	for _, rule := range cfg.ExternalRules {
+		if externalConverter, err := converterFromRule(rule, mapper); err == nil {
+			converters = append(converters, externalConverter)
+		} else {
+			errs = append(errs, err)
+		}
+	}
+	return converters, errs
+}
+
+func converterFromRule(rule config.DiscoveryRule, mapper apimeta.RESTMapper) (SeriesConverter, error) {
+	var (
+		err error
+	)
+
+	resourceConverter, err := naming.NewResourceConverter(rule.Resources.Template, rule.Resources.Overrides, mapper)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create ResourceConverter associated with series query %q: %v", rule.SeriesQuery, err)
+	}
+
+	queryBuilder, err := NewQueryBuilder(rule.MetricsQuery)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create a QueryBuilder associated with series query %q: %v", rule.SeriesQuery, err)
+	}
+
+	seriesFilterer, err := NewSeriesFilterer(rule.SeriesFilters)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create a SeriesFilter associated with series query %q: %v", rule.SeriesQuery, err)
+	}
+
+	if rule.Name.Matches != "" {
+		err := seriesFilterer.AddRequirement(config.RegexFilter{Is: rule.Name.Matches})
+		if err != nil {
+			return nil, fmt.Errorf("unable to apply the series name filter from name rules associated with series query %q: %v", rule.SeriesQuery, err)
+		}
+	}
+
+	metricNamer, err := naming.NewMetricNamer(rule.Name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create a MetricNamer associated with series query %q: %v", rule.SeriesQuery, err)
+	}
+
+	return &seriesConverter{
+		seriesQuery:       prom.Selector(rule.SeriesQuery),
+		mapper:            mapper,
+		resourceConverter: resourceConverter,
+		queryBuilder:      queryBuilder,
+		seriesFilterer:    seriesFilterer,
+		metricNamer:       metricNamer,
+	}, nil
+}
+
+func (c *seriesConverter) buildNamespaceQueryPartForExternalSeries(namespace string) (queryPart, error) {
+	namespaceLbl, _ := c.metricNamer.LabelForResource(naming.NsGroupResource)
```

Comment:
`c.metricNamer.LabelForResource` will fail here due to nil pointer dereferencing, it should be `c.resourceConverter.LabelForResource`